meeko/molsetup.py

```diff
@@ -4,623 +4,451 @@
-        molsetup.perceive_rings(keep_chorded_rings, keep_equivalent_rings)
-        return molsetup
+    def from_prmtop_inpcrd(cls, prmtop, crd_filename: str):
+        # TODO: pull functionality over and clean up
+        pass
 
-    @classmethod
-    def from_prmtop_inpcrd(cls, prmtop, crd_filename):
-        molsetup = cls()
-        x, y, z = prmtop._coords_from_inpcrd(crd_filename)
-        all_atom_indices = prmtop.vmdsel("")
-        if len(all_atom_indices) != len(x):
-            raise RuntimeError("number of prmtop atoms differs from x coordinates")
-        prmtop.upper_atom_types = prmtop._gen_gaff_uppercase()
-        charges = [chrg/18.2223 for chrg in prmtop.read_flag("CHARGE")]
-        atomic_nrs = prmtop.read_flag("ATOMIC_NUMBER")
-        pdbqt_string = ""
-        molsetup.atom_params["rmin_half"] = []
-        molsetup.atom_params["epsilon"] = []
-        vdw_by_atype = {}
-        for vdw in prmtop._retrieve_vdw():
-            if vdw.atom in vdw_by_atype:
-                raise RuntimeError("repeated atom type from prmtop._retrieve_vdw()")
-            vdw_by_atype[vdw.atom] = {"rmin_half": vdw.r, "epsilon": vdw.e}
-        for i in prmtop.atom_sel_idx:
-            pdbinfo = rdkitutils.PDBAtomInfo(
-                name=prmtop.pdb_atom_name[i],
-                resName = prmtop.pdb_resname[i],
-                resNum = int(prmtop.pdb_resid[i]),
-                chain = " ",
-            )
-            atype = prmtop.upper_atom_types[i]
-            molsetup.add_atom(
-                i,
-                coord = (x[i], y[i], z[i]),
-                charge = charges[i],
-                atom_type = atype,
-                element = atomic_nrs[i], 
-                pdbinfo = pdbinfo,
-                chiral = False,
-                ignore = False,
-                add_atom_params = {
-                    "rmin_half": vdw_by_atype[atype]["rmin_half"],
-                    "epsilon": vdw_by_atype[atype]["epsilon"],
-                }
+    # region Manually Building A MoleculeSetup
+    def add_atom(
+        self,
+        atom_index: int = None,
+        overwrite: bool = False,
+        pdbinfo: str = DEFAULT_PDBINFO,
+        charge: float = DEFAULT_CHARGE,
+        element: int = DEFAULT_ELEMENT,
+        atom_type: str = DEFAULT_ATOM_TYPE,
+        is_ignore: bool = DEFAULT_IS_IGNORE,
+        is_chiral: bool = DEFAULT_IS_CHIRAL,
+        graph: list[int] = field(default_factory=list),
+    ):
+        """
+        Adds an atom with all the specified attributes to the MoleculeSetup, either at the specified atom index, or by
+        appending it to the internal list of atoms. Default values will be used for any attributes with unspecified
+        values.
+
+        Parameters
+        ----------
+        atom_index: int
+        overwrite: bool
+        pdbinfo: str
+        charge: float
+        element: int
+        atom_type: str
+        is_ignore: bool
+        is_chiral: bool
+        graph: List[List[int]]
```

Comment:
There seem to be some implementations in the original molecule setup that use graph to refer to the overall attribute, but neighbors to refer to the connected nodes specific to an atom. Would it be more prudent to just call the attribute neighbors?

Reply:
Not sure I fully understood, but I think the answer is yes. Each atom should know about it's immediate neighbors, not the neighbors of every other atom. I see the full graph at the level of the entire molecule, not at the level of each atom. Moreover, the graph is really just the bonds, it duplicates the same information that the bonds already have. I think we should remove the graph, and the methods that currently use the graph would instead use the bonds.

Reply:
The neighbors are also redundant with bonds, but maybe it's convenient to keep both.